dpctl/apis/include/dpctl4pybind11.hpp

```diff
@@ -748,6 +748,64 @@ class usm_memory : public py::object
             throw py::error_already_set();
     }
 
+    /*! @brief Create usm_memory object from shared pointer that manages
+     *  lifetime of the USM allocation.
+     */
+    usm_memory(void *usm_ptr,
+               size_t nbytes,
+               const sycl::queue &q,
+               std::shared_ptr<void> shptr)
+    {
+        auto const &api = ::dpctl::detail::dpctl_capi::get();
+        DPCTLSyclUSMRef usm_ref = reinterpret_cast<DPCTLSyclUSMRef>(usm_ptr);
+        sycl::queue *q_ptr = new sycl::queue(q);
+        if (!q_ptr) {
+            throw std::bad_alloc();
+        }
+        DPCTLSyclQueueRef QRef = reinterpret_cast<DPCTLSyclQueueRef>(q_ptr);
+
+        auto vacuous_destructor = []() {};
+        py::object mock_owner = py::capsule(vacuous_destructor);
+
+        // create memory object owned by mock_owner, it is a new reference
+        PyObject *_memory =
+            api.Memory_Make_(usm_ref, nbytes, QRef, mock_owner.ptr());
+
+        bool is_ok =
+            (PyObject_TypeCheck(
+                 _memory, ::dpctl::detail::dpctl_capi::get().Py_MemoryType_) !=
+             0);
+
+        if (is_ok) {
+            std::shared_ptr<void> *opaque_ptr =
+                new std::shared_ptr<void>(shptr);
+            if (!opaque_ptr) {
```

Comment:
If I'm not mistaken `new` throws `std::bad_alloc` instead of returning `nullptr` in case it fails to allocate memory.
Unless you specified `std::nothrow`

https://en.cppreference.com/w/cpp/memory/new/nothrow

So you wouldn't reach that point and wouldn't be able to free memory unless you are using `unique_ptr`. Which is not very important, since if we are out of memory application would crash anyway.

Reply:
I changed the code to expect `new` to throw an exception. Please have another look